src/urbit/lib/chess.hoon

```diff
@@ -399,13 +407,107 @@
       ==
   --
 ::
-::  XX: implement or delete
+::  XX: need to implement
+::  turn fen into chess-position with checks for valid fen
+++  fen-to-position-safe
+|=  fen=@t
+~&  %not-implemented  !!
 ::
-::  would read position from fen,
-::  but hasn't been implemented
+::  turn fen into chess-position
 ++  fen-to-position
-  |=  fen=@t
-  ~&  %not-implemented  !!
+|=  fen=@t
+^-  chess-position
+%+  rush
+  fen
+|.
+;~  (glue ace)
+    fen-to-board         ::  board
+    fen-to-player        ::  player-to-move
+    fen-to-white-castle  ::  white-can-castle
+    fen-to-black-castle  ::  black-can-castle
+    fen-to-en-passant    ::  en-passant
+    dem                  ::  ply-50-move-rule
+    dem                  ::  move-number
+==
+++  fen-to-board
+  %+  cook  board-helper
+  %+  more  (just '/')
+  ::  XX is this actually slower than alf and fail on invalid char?
+  (stun [1 8] (mask "12345678bknpqrBKNPQR"))
+++  fen-to-player
+  ;~  pose
+      (cold %white (just 'w'))
+      (cold %black (just 'b'))
+  ==
+++  fen-to-white-castle
+  ;~  pose
+      (white-castle-helper hep " -" %none)
+      (white-castle-helper (jest 'k') " k" %none)
+      (white-castle-helper (jest 'q') " q" %none)
+      (white-castle-helper (jest 'KQ') " " %both)
+      (white-castle-helper (just 'K') " " %kingside)
+      (white-castle-helper (just 'Q') " " %queenside)
+  ==
+++  fen-to-black-castle
+  ;~  pose
+      (cold %none hep)
+      (cold %both (jest 'kq'))
+      (cold %kingside (just 'k'))
+      (cold %queenside (just 'q'))
+  ==
+++  fen-to-en-passant
+  ;~  pose
+      (cold ~ hep)
+      %+  cook
+          some
+          ;~  plug
+              (cook term (shim 'a' 'h'))
+              (cook term (mask "36"))
+          ==
+  ==
+++  board-helper
+  |=  brd=(list tape)
+  ?>  =((lent brd) 8)
+  =/  rank=chess-rank  %8
+  =/  file=chess-file  %a
+  =/  board  *(map chess-square chess-piece)
+  |-
+  ?~  brd
+    board
+  ?<  (gth file %g)
+  ?:  =(file %g)
+    $(brd t.brd, rank (need (prev-rank rank)), file %a)
+  =*  next  (snag (sub file 97) i.brd)
```

Comment:
Okay, I made some mistakes in my first draft (including forgetting that `%g` comes *before* `%h`, not after it...). Try this:
```
=/  rank=chess-rank  %8
=/  file=$?(chess-file %i)  %a
...
  board
?-  file
    %i
  $(brd t.brd, rank (chess-rank (sub rank 1)), file %a)
::
    chess-file
  =*  next  (snag (sub file 97) i.brd)
  ...
==
```